cuda/matrix/ell_kernels.cu

```diff
@@ -128,16 +166,97 @@ void spmv(std::shared_ptr<const CudaExecutor> exec,
+    const dim3 grid_size(
+        ceildiv(nrows * subwarp_size * nwarps_per_row, block_size.x),
+        b->get_size()[1], 1);
+    switch (subwarp_size) {
+    case 1:
+        abstract_spmv<1><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 2:
+        abstract_spmv<2><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 4:
+        abstract_spmv<4><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 8:
+        abstract_spmv<8><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 16:
+        abstract_spmv<16><<<grid_size, block_size, 0, 0>>>(
+            nrows, as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+            a->get_stride(), a->get_num_stored_elements_per_row(),
+            as_cuda_type(b->get_const_values()), b->get_stride(),
+            as_cuda_type(c->get_values()), c->get_stride());
+        break;
+    case 32:
+        if (atomic) {
+            ASSERT_NO_CUDA_ERRORS(
+                cudaMemset(c->get_values(), 0,
+                           c->get_num_stored_elements() * sizeof(ValueType)));
+            abstract_spmv<32, true><<<grid_size, block_size, 0, 0>>>(
+                nrows, as_cuda_type(a->get_const_values()),
+                a->get_const_col_idxs(), a->get_stride(),
+                a->get_num_stored_elements_per_row(),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(c->get_values()), c->get_stride());
+        } else {
+            abstract_spmv<32><<<grid_size, block_size, 0, 0>>>(
+                nrows, as_cuda_type(a->get_const_values()),
+                a->get_const_col_idxs(), a->get_stride(),
+                a->get_num_stored_elements_per_row(),
+                as_cuda_type(b->get_const_values()), b->get_stride(),
+                as_cuda_type(c->get_values()), c->get_stride());
+        }
```

Comment:
I really want all of this to go away. I'm trying to see what can be done for that. In general, because the possible values are well known beforehand, we should generate at compile time the kernel for the versions with all possible subwarp sizes (1, 2, 4, 8, 16, 32) and atomicity (true, false). I think we could go with a solution similar to what Goran used for the block jacobi. I well tell you more details when I have them.

Reply:
Here is an example of using Goran's code for selecting the subwarp size:
https://pastebin.com/qwyWUAPX

The code can still be cleaned, I only did that quickly so it's far from perfect. I also did not do the advanced_apply there.

Finally, it should be possible to do something about the `atomic` to remove the final test.